### PR TITLE
Fix GAN structure

### PR DESCRIPTION
**Pull request by amjltc295**
## Why do we need this PR?
* Make optimizers and lr_scheduler dictionaries

## How is it implemented?
*

### CODE REVIEW

File: src/pipeline/base_pipeline.py [hunk 1]
```diff
@@ -203,11 +203,11 @@ def _resume_checkpoint(self, resumed_checkpoint):
...
             logger.warning('Warning: Optimizer type given in config file is different from that of resumed_checkpoint. '
                            'Optimizer parameters not being resumed.')
-        elif self.optimizer is None:
+        elif self.optimizers is None:
             logger.warning("Not loading optimizer state because it's not initialized.")
         else:
-            for key, optimizer_state in resumed_checkpoint['optimizer'].items():
-                self.optimizer[key].load_state_dict(optimizer_state)
+            for key, optimizer_state in resumed_checkpoint['optimizers'].items():
+                self._setup_optimizers[key].load_state_dict(optimizer_state)
```
Comment (Nash2325138) [inline src/pipeline/base_pipeline.py]: self.optimizers instead of self._setup_optimizers?

File: src/model/loss.py [hunk 2]
```diff
@@ -55,3 +105,48 @@ def forward(self, data_dict, output_dict):
         loss_real = self.relu(1 - D_x).mean()
         loss_fake = self.relu(1 + D_G_z).mean()
         return loss_real + loss_fake
+
+
+# Based on https://github.com/knazeri/edge-connect/blob/master/src/loss.py
+class AdversarialLoss(nn.Module):
```
Comment (Nash2325138) [inline src/model/loss.py]: This class is not used and should be removed

File: src/model/loss.py [hunk 3]
```diff
@@ -55,3 +105,48 @@ def forward(self, data_dict, output_dict):
...
+        self.register_buffer('real_label', torch.tensor(target_real_label))
+        self.register_buffer('fake_label', torch.tensor(target_fake_label))
+
+        if type == 'nsgan':
+            self.criterion = nn.BCELoss()
+
+        elif type == 'lsgan':
+            self.criterion = nn.MSELoss()
+
+        elif type == 'hinge':
```
Comment (Nash2325138) [inline src/model/loss.py]: hinge loss is not supported according to `__call__`

File: src/data_loader/mnist.py [hunk 4]
```diff
@@ -15,3 +16,21 @@ def __getitem__(self, index):
...
+    """
+    Customized MNIST dataset demo
+    """
+    def __init__(self, data_dir, train, download, transform):
+        super().__init__(data_dir, train=train, download=download, transform=transform)
+
+    def __getitem__(self, index):
+        """ Overwrite __getitem__ to return dictionary """
+        data = super().__getitem__(index)
+        z = torch.randn((1, 100)).view(100, 1, 1)
```
Comment (Nash2325138) [inline src/data_loader/mnist.py]: Why is `z` now initialized on data loading?

File: src/data_loader/mnist.py [hunk 5]
```diff
@@ -15,3 +16,21 @@ def __getitem__(self, index):
             "data_input": data[0],
             "data_target": data[1]
         }
+
+
+class MnistGANDataset(datasets.MNIST):
```
Comment (Nash2325138) [inline src/data_loader/mnist.py]: This class is never used

File: src/worker/validator.py [hunk 6]
```diff
@@ -12,16 +12,28 @@ class Validator(TrainingWorker):
     def __init__(self, pipeline: BasePipeline, *args):
         super().__init__(pipeline, *args)
         # Some shared attributes are validator exclusive and therefore is initialized here
-        for attr_name in ['loss_functions']:
+        for attr_name in ['optimizers', 'loss_functions', 'optimize_strategy']:
```
Comment (Nash2325138) [inline src/worker/validator.py]: `optimizers` should not be available for validator

File: src/configs/template_GAN_train_config.json [hunk 7]
```diff
@@ -57,18 +55,22 @@
...
     },
     "losses": {
-        "0": {
-            "type": "HingeLossG",
+    },
+    "gan_losses": {
+        "generator": {
+            "type": "GANLoss",
             "args": {
                 "nickname": "hinge_loss_generator",
```
Comment (Nash2325138) [inline src/configs/template_GAN_train_config.json]: By default values of class `GANLoss` constructor, the nickname should be `LSGAN_loss_genrator`

File: src/configs/template_GAN_train_config.json [hunk 8]
```diff
@@ -57,18 +55,22 @@
...
+                "weight": 1,
+                "network": "generator"
             }
         },
-        "1": {
-            "type": "HingeLossD",
+        "discriminator": {
+            "type": "GANLoss",
             "args": {
                 "nickname": "hinge_loss_discriminator",
```
Comment (Nash2325138) [inline src/configs/template_GAN_train_config.json]: The nickname should be `LSGAN_loss_discriminator`
